KEP-4671 : Gang Scheduling (#5558)
kubernetes/enhancements: keps/sig-scheduling/4671-gang-scheduling/README.md

    // This field is immutable.
    ControllerRef *v1.ObjectReference

    // PodGroups is a list of groups of pods.
    // Each group may request gang scheduling.
    PodGroups []PodGroup 
}

type GangMode string
const (
	// GangModeOff means that all pods in this PodGroup do not need to be scheduled as a gang.
	GangModeOff GangMode = "Off"

	// GangModeSingle means that all pods in this PodGroup need to be scheduled as one gang.
	GangModeSingle GangMode = "Single"

	// GangModeReplicatedGang means that there is a variable number of identical copies of this PodGroup,
    //  as specified in Replicas, and each copy needs to be independently gang scheduled.
	GangModeReplicated GangMode = "Replicated"
)

// GangSchedulingPolicy holds options that affect how gang scheduling of one PodGroup is handled by the scheduler.
type GangSchedulingPolicy struct {
    // SchedulingTimeoutSeconds defines the timeout for the scheduling logic.
    // Namely it's timeout from the moment when `minCount` pods show up in
    // PreEnqueue, until those pods are observed in WaitOnPermit - for context
    // see https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#interfaces

> 	SchedulingTimeoutSeconds *int

Reviewer: - What if it reaches `SchedulingTimeoutSeconds`? I guess you meant to reject all waiting pods and try scheduling them from scratch once again, but you need a clarification.
- Is it really supposed to be exposed to users? 
  - Users usually don't have a good understanding of the cluster's scheduling throughput etc. Isn't it hard for users to determine it in the first place? 
  - From the cluster admin pov as well, if some users have really long timeout, their pods can keep resources for a long time, blocking other pods.
- What happens if we move to a new phase at Beta? Like you described in `For Beta, we want to also touch ...` in "Scheduler Changes" section, we'll reserve a resource before starting schedule of individual pods. Will this `SchedulingTimeoutSeconds` still make sense after Beta? given we will "compute the placement for all of these pods in a single scheduling cycle" (quote your explanation) ?

Reviewer: > That being said - I'm happy to be challenged here - this is not a core and if we're not convinced we can remove it from here and consider adding that in a follow-up (and for now make it a scheduler value).

Yeah, I think we can just move it to a scheduler config for now and consider it in the future, if we find a good motivation to expose it to users.
But, I need to be convinced this knob is necessary even after Beta ↓

> This is still fine. Because even though we propose a placement for all pods in a single cycle, the individual pods will later go through the existing phases on their own (to validate the placement, etc.). So that timeout still technically makes sense.

I imagined if some pods are rejected (e.g., high priority pod stealed the place etc), we always have to re-compute all pods' placement from scratch, rather than keep some pods in WaitOnPermit and keep others trying to get the place. Otherwise I thought the deadlock problem won't be solved?

Reviewer: > Right - maybe I explained it poorly, but what you wrote is definitely what we need to do.

Then, will SchedulingTimeoutSeconds no longer make sense after Beta? Because we won't keep pods in WaitOnPermit anymore.

Author: > Then, will SchedulingTimeoutSeconds no longer make sense after Beta? Because we won't keep pods in WaitOnPermit anymore in case of some pods' unschedulable results.

We want to still keep them in WaitOnPermit, because the assignment computed in a single phase may appear no longer valid in the meantime. Because the pods come from this single phase only with "nomination" which is then validated using regular pod-by-pod scheduling. So it still makes sense.

But I generally buy the argument that it's unclear to majority of users how to set this value, so we probably don't need it from the very beginning.

Reviewer 2: I think that originally idea behind this timeout was to define a time how long scheduler should try to find a feasible placement. It was rather supposed to be a part of the workload scheduling phase, so before entering pod-by-pod scheduling.

We reduced it to the pod-by-pod WaitOnPermit, which I agree, it's a technical detail how long we expect pods to be scheduled. It's technical although important, as some autoscaler needs to have enough time to notice unschedulable pods and provision necessary nodes within the timeout.

> I imagined if some pods are rejected (e.g., high priority pod stealed the place etc), we always have to re-compute all pods' placement from scratch

Yes, and this is where the new workload scheduling phase is important. If WaitOnPermit timed out, it means that the placement it prepared is no longer valid (it turned out to be not working anymore for some reason) and it's a signal that a new placement needs to be proposed.

Reviewer 2: > But, as I https://github.com/kubernetes/enhancements/pull/5558#discussion_r2400975499, if a placement is no longer valid, we need to re-compute all pods' placements, no? We won't keep some pods in WaitOnPermit while keeping re-scheduling other pods because that'd result in the deadlock scenario.

We can always reject pods on WaitOnPermit in situations like you said before the timeout expires, but I think we should keep the timeout even then, for instance to let autoscaler provides new nodes.

Yes, we can move waiting on nodes provisioning after WaitOnPermit in the future (maybe even for beta), but assuming that workload scheduling will prepare pods placement and assign them to non existing nodes (as in our original NNN KEP). If not, autoscaling needs to provision nodes before WaitOnPermit expires.

Author: There are two angles of this discussion:
1) Whether this timeout is needed at all
2) Whether this timeout should be exposed via workload API.

The core of this discussion is (2) - we can still have the timeout (I personally agree we need this timeout, although even here we don't seem to be aligned). 
But the fact that we need a timeout doesn't mean it needs to be immediately exposed directly in the `Workload` API. We can start with it being hardcoded, or be exposed as scheduler config parameter so still give a chance to tune it for the most sophisticated operators, without putting it directly to the API from the very beginning.

And by removing it from here, we didn't say we won't have this timeout - we only said it's not part of the API yet.

Reviewer 3: Supportive of moving timeout from the spec into scheduler config for alpha.  This probably needs some tuning at the cluster level, but letting individual jobs tune it could deny service to other jobs.

 A new core type called `Workload` is introduced to tell the kube-scheduler that a group of pods should be scheduled together and any policy options related to gang scheduling. Pods have an object reference in their spec to their `Workload`, if any. The `Workload` object is intended to evolve[^2] via future KEPs to support additional kube-scheduler improvements, such as topology-aware scheduling,  

## Motivation

Parallel applications can require communication between every pod in order to begin execution, and then ongoing communication between all pods (such as barrier or all-reduce operations) in order to make progress.  Starting all pods at close to the same time is necessary to run these workloads.  Otherwise, either expensive compute resources are idle, or the application may fail due to an application-level communication timeout.

> Gang scheduling has been implemented outside of kube-scheduler at least 4 times[^3].  Some controllers are starting to support multiple Gang Schedulers in order to be portable across different clusters.  Moving support into kube-scheduler makes gang scheduling support available in all Kubernetes distributions and eventually may allow workload controllers to reply on a standard interface to request gang scheduling from the standard or custom schedulers. A standard API may also allow other components to understand workload needs better (such as cluster autoscalers).

Reviewer 4: Thank you. The comparison mentioned in https://docs.google.com/document/d/1ulO5eUnAsBWzqJdk_o5L-qdq5DIVwGcE7gWzCQ80SCM/edit?pli=1&tab=t.3zjbiyx2yldg focuses on the API aspect of the gang scheduling. Not their lifecycle. Perhaps could the paragraph mention the lessons learnt here are focusing on the API and the actual implementation (including the lifecycle) will keep adapting through time?

The Workload object will allow kube-scheduler to be more aware that pods are part of workloads with complex internal structure.  Those workloads include builtins like `Job` and `StatefulSet`, and custom workloads, like `JobSet`, `LeaderWorkerSet` and `MPIJob`. All of these workload types are used for AI training and inference use cases.  

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> ### Goals

Reviewer: In the doc, I saw:

>  it’s not necessary to introduce sophisticated mechanisms preventing deadlocks in the pod-by-pod scheduling process, as it’s expected that pods entering the scheduling cycle already have precomputed placement inside an atomically placed reservation. In such a case, all pods released to scheduling will be expected to schedule successfully in one pass or invalidate the plan and trigger its rescheduling along with its resource reservation.

This is a very very crucial factor, determining how much we should refine this feature, with solving the known problems like I mentioned in another comment.

If we'll argue that this feature is not intended to use alone, but is to be used with Kueue etc that ensures the capacity to some extent, outside the scheduler, I'd agree that we can loosen the quality of gang scheduling implementation. But, we'd need to clearly deliver that intention on the doc and anywhere else.

Reviewer: This proposal looks like some enhancements but based on what the existing coscheduling plugin does. 
Honestly, when I saw this for the first time just an hour ago, it looked not great because it's leaving some obviously-known/already-raised issue unsolved. But, the above explanation can be a big changer of my mind. So, again, I believe that is very crucial point that we need to declare at the very first of this KEP and the documentation as well.

My core thought is: we shouldn't put an effort into a huge break into the current scheduling fundamental layer, while knowing they'll be wasted: like there'll be a long term WAS solution that will replace all those changes completely. In that case, investing everyone's time into this short-term solution would be just a delay of a long-term proper gang implementation.

So, I'd like to make sure that we won't go too-different direction as a short term solution. Ideally, what we'll implement here should be a subset of a future WAS change: e.g., a future gang scheduling through WAS framework would be eventually going through this gang plugin, after some fancy calculation etc outside the current scheduling cycle.

Author: @sanposhiho @dom4ha ^^

Reviewer 2: >  In that case, investing everyone's time into this short-term solution would be just a delay of a long-term proper gang implementation.

It's definitely not a short-term change, but indeed, we haven't provided anywhere a bigger picture yet. The idea is that the pod-by-pod scheduling stays with us forever, so as Wojtek said, we should distinguish two (or even three) levels of scheduling where pod-by-pod is the lowest one:

1. pod-by-pod - the scheduler does individual pod scheduling unless it receives a proposed or mandatory placement from the 2nd (or 3rd) level scheduling. In that case it does mainly correctness validation, resources tracking and workload-aware binding. The proposal in this KEP is only about this low level scheduling-binding, but it's reduced to the very minimum.
2. workload scheduling - at this level we need to have ability to schedule multiple pods at once, or even find a placement for a workload which does not have pods created yet (like scheduling future reservations). We believe that the current kube-scheduling framework is not suitable to do that, because you cannot reason when cluster state constantly changes between processing consecutive pods etc. So in the future extensions (not for 1.35 due to time constraints only) we will propose introducing an new atomic phase which will compute a placement for pods, so the actual pods going though low level scheduling should have their mandatory placement precomputed (for now we can use NNN, but it may potentially change). It's important to note that whenever the execution of that plan fails on the pod-by-pod layer, a new workload placement needs to be created. Another important note is that kube-scheduler would be open on external workload schedulers which could propose a workload placement which kube-scheduler does not need to understand-why, just validate whether it is correct and whether it can successfully apply it.
3. orchestration layer - some clusters needs optimization that cannot be achieved by scheduling workload-by-workload, but needs to consider multiple workloads at once and consider their timing aspects. So Kueue is a good example of such scheduler as it can reason on a higher level when it makes sense to start a workload and which workloads should be preempted. So the workload placement proposed at this level can be consumed by workload scheduler at the lower level without a need to understand-why as long as it's correct. If it turns out to be not correct (cluster state has changed in the meantime), we may come back from pod-by-pod level up to orchestration level asking for recomputing a placement of many workloads.

Note that the WaitOnPermit gate (or to be precise, a similar one that we may want to define) will serve more than a gang-scheduling gate on this picture. Let's call it WaitOnReschedule which will go after WaitOnPermit and before WaitOnPreemption (when the preemption is physically applied on pods). We should be able to schedule a few workloads (coming from the orchestration layer) that pass WaitOnPermit and verify that all of them have an accepted placement on WaitOnReschedule before we let such multi-workload plan to be executed (allow to perform preemption) on WaitOnPreemption. There may be more such gates, but they are out of scope in this KEP.

I know it deserves a separate doc, so apologies we haven't published too many docs yet.

Reviewer: OK, let me clarify my points. I'm not disagreeing with this KEP as a whole, first of all. 

My main points are:
1. Going for beta with this design is not OK, especially if you're foreseeing the future that users should use Workload API directly, not via Kueue or whatever solution on top of K8s. We have many known problems and (at least from this KEP) I don't see any solution. 
2. But, going for alpha with a simple implementation does make sense. Especially, the sooner we establish the API, the better for adopters (Kueue etc), even if the internal implementation is rough at that point.
3. However, it's also not OK if we're going for alpha with a simple implementation that does not match with our future direction. This one worried me because, as you both also mentioned, I'm pretty sure that those problem solving would necessary a whole new mechanism on top of what we have (yes let's not talk about the detail of machine gun here), not just an extension of what we have, like the current KEP tries to implement.

So, basically, I'm ok with this KEP going into this release, as long as my concerns (1) (3) are taken into consideration.

Based on them, my requests to make sure those on this KEP is:
1. Please list up the known problems. Otherwise it's hard to talk what will focus in the alpha stage, and what we'll carry over to the beta stage with a whole new mechanism.
2. No need to talk too much / finalize with this release cycle, but please put an overall picture of a whole new mechanism so that we can _roughly_ make sure we're not going away from the future design.

Reviewer: ...and I just saw some commits yesterday. Let me check now.

    // Each value of that key maps to a different identical gang, which will be anonymous in status,
    // and referred to by the key's value in events and errors.
    ReplicatedGangKey *ObjectFieldSelector

    // Optional when GangMode = "ReplicatedGang".
    // Forbidden otherwise.
    Replicas int

    // PodGroupSelector must be set when GangMode = "Single" or "Off".
    // Must be one of:
    //   - `metadata.labels['<KEY>']`
    //   Note: Annotation not supported because it is not indexed easily, and it may not
    //         have a form that can be printed without escaping.
    // Each value of that key maps to a different gang, which will be anonymous in status,
    // and referred to by the key's value in events and errors.
    PodGroupSelector PodGroupSelector

    // GangSchedulingPolicy defines the options applying to all pods in this gang.
    // Forbidden if GangMode is set to "Off".
    GangSchedulingPolicy GangSchedulingPolicy
}

type WorkloadStatus struct {
  // Necessary status fields TBD.
}
```

### Scheduler Changes

The Gang Scheduling functionality will be implemented using a combination of kube-scheduler plugins and scheduling framework changes.

> The plugin will use  PreEnqueue and WaitOnPermit framework hooks to control the scheduling process of a gang. The scheduling framework changes will watch Workloads, map pods to and from their Workloads and Gang(s) within the workload. 

Author: done

> More detail about scheduled changes is described in [this document](https://docs.google.com/document/d/1lMYkDuGqEoZWfE2b8vjQx0vHieOMyfmi6VHUef5-5is/edit?tab=t.0#heading=h.1p88ilpefnb).

Author: In particular, feel free to tag me on certain parts of that doc and I'll try to incorporate them here.

> More detail about scheduled changes is described in [this document](https://docs.google.com/document/d/1lMYkDuGqEoZWfE2b8vjQx0vHieOMyfmi6VHUef5-5is/edit?tab=t.0#heading=h.1p88ilpefnb).

Reviewer: I already saw some addition here, thanks.

> Every single KEP that I've seen mentioning too many implementation details inside was was soon after diverging from the implementation.

Yeah, like implementation details on each extension point is not necessary because, like you said, they would just be too detailed to discuss on KEP. I think everyone can get a rough grasp of what will be implemented with your `for Alpha, we will implement...` explanation.

But, (as I just mentioned on the other thread though) I'd like to see a list of problems/features that we eventually need to solve/implement, such as a deadlock problem, a better requeueing mechanism (anything around queueing like the ordering in activeQ), workload-aware preemption, ..maybe something else?
I saw discussion for some of them on the google doc side, and I do think we need them mentioned here so that we can discuss a rough picture of solution and what we should do at alpha/beta, and also make sure we're not missing some problems at this point.

> More detail about scheduled changes is described in [this document](https://docs.google.com/document/d/1lMYkDuGqEoZWfE2b8vjQx0vHieOMyfmi6VHUef5-5is/edit?tab=t.0#heading=h.1p88ilpefnb).

Author: The high-level nature of this comment has already been addressed -  we're still discussing more targeted comments around that, so resolving his one.

Status updates:
  - API call type: PUT/PATCH Workloads
  - estimated throughput < XX/s
  - originating component: kube-scheduler

###### Will enabling / using this feature result in introducing new API types?

Yes:
  - API type: Workload
  - Supported number of objects per cluster: XX,000
  - Supported number of objects per namespace: XX,000

The above numbers should eventually match the numbers for built-in workload APIs
(e.g. Deployments, Jobs, StatefulSets, ...).

###### Will enabling / using this feature result in any new calls to the cloud provider?

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

Yes. New field (spec.workload) is added to the Pod API:
  - API type: Pod
  - Estimated increase in size: XX-XXX bytes per object (depending on the final choice described
    in the Associating Pod into PodGroups section above).

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

Pod startup SLI/SLO may be affected and should be adjusted appropriately.
The reason is that scheduling a pod being part of a gang will now be blocked on all pods

> from a gan to be created and observed by the scheduler (which from large gangs can take

Author: Done

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> In this KEP, kube-scheduler is modified to support gang scheduling[^1]. To implement gang scheduling, kube-scheduler identifies pods that are in a group and waits until all pods reach the same stage of the scheduling/binding cycle before allowing any pods from the group to advance past that point.  If not all pods can reach that point before a timeout expires, then the scheduler stops trying to schedule that group, and all pods release all their resources.  This allows other workloads to try to allocate those resources.

Reviewer 2: > In this KEP, kube-scheduler is modified to support gang scheduling[^1]. To implement gang scheduling, kube-scheduler identifies pods that are in a group

I think we should make it clear from the very beginning, that this particular KEP is not necessarily implementing scheduling of gangs, but supports Gang Scheduling as stated in the first sentence, or is the first step toward implementing GS in the kube-scheduler. The focus of this KEP Is mainly on defining API and we could make it clear from the very beginning,

A new core type called `Workload` is introduced to tell the kube-scheduler that a group of pods should be scheduled together and any policy options related to gang scheduling. Pods have an object reference in their spec to their `Workload`, if any. The `Workload` object is intended to evolve[^2] via future KEPs to support additional kube-scheduler improvements, such as topology-aware scheduling.

## Motivation

Parallel applications can require communication between every pod in order to begin execution, and then ongoing communication between all pods (such as barrier or all-reduce operations) in order to make progress.  Starting all pods at close to the same time is necessary to run these workloads.  Otherwise, either expensive compute resources are idle, or the application may fail due to an application-level communication timeout.

> Gang scheduling has been implemented outside of kube-scheduler at least 4 times[^3].  Some controllers are starting to support multiple Gang Schedulers in order to be portable across different clusters.  Moving support into kube-scheduler makes gang scheduling support available in all Kubernetes distributions and eventually may allow workload controllers to reply on a standard interface to request gang scheduling from the standard or custom schedulers. A standard API may also allow other components to understand workload needs better (such as cluster autoscalers).

Reviewer 2: ```suggestion
Gang scheduling has been implemented outside of kube-scheduler at least 4 times[^3].  Some controllers are starting to support multiple Gang Schedulers in order to be portable across different clusters.  Moving support into kube-scheduler makes gang scheduling support available in all Kubernetes distributions and eventually may allow workload controllers to rely on a standard interface to request gang scheduling from the standard or custom schedulers. A standard API may also allow other components to understand workload needs better (such as cluster autoscalers).
```

## Design Details

### Naming

* `Workload` is the resource Kind.
* `scheduling` is the ApiGroup.
* `spec.workload` is the name of the new field in pod.
* Within a Workload there is a list of groups of pods. Each group represents a top-level division of pods within a Workload.  Each group can be independently gang scheduled (or not use gang scheduling). This group is named
  <<[UNRESOLVED community feedback requested]>> `PodGroup` or `GangGroup` for the top level. <<[/UNRESOLVED]>>.
* In a future , we expect that this group can optionally specify further subdivision into sub groups.  Each sub-group can have an index.  The indexes go from 0 to N, without repeats or gaps. These subgroups are called
  <<[UNRESOLVED depending on previous unresolved item]>> `PodSubGroup` if `PodGroup` is chosen, or else `RankedGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.
* In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called

>   <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.

Author: It seems we're going with PodGroup anyway (GangGroup doesn't make sense for non-gang option), so I just went ahead and removed all of these unresolved comments.

// all list items.  Every list item must have a distintict value for this key.  This ensures
// pod groups have disjoint sets of pods in them.
type PodGroupSelector struct {
    // Selector that identifies pods in this gang.
    Selector *metav1.LabelSelector
}

// PodGroup is a group of pods that may contain multiple shapes (EqGroups) and may contain
// multiple dense indexes (RankedGroups) and which can optionally be replicated in a variable
// number of identical copies.
//
// TODO: Decide on the naming: PodGroup vs GangGroup.
type PodGroup struct {
    Name *string
    GangMode *GangMode // default is "Single"

	// ReplicatedGangKey must be set when GangMode = "Replicated"
    // It must be one of:
    //   - `metadata.labels['<KEY>']`
    //   Note: Annotations are not supported because they are not indexed easily, and may not
    //         have a form that can be printed without escaping.
    // Each value of that key maps to a different identical gang, which will be anonymous in status,
    // and referred to by the key's value in events and errors.
    ReplicatedGangKey *ObjectFieldSelector

    // Optional when GangMode = "ReplicatedGang".

>     // Forbidden otherwise.

Author: OK - given the overall consensus (@sanposhiho @andreyvelich @macsko myself) about this version, I just went ahead I made this the leading option and moved the original option down to alternatives.

>     // Forbidden otherwise.

Reviewer 5: Looks great and similar to [the VolumeSource API](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L44).
Thanks!

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) within one minor version of promotion to GA
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> In this KEP, kube-scheduler is modified to support gang scheduling[^1]. To implement gang scheduling, kube-scheduler identifies pods that are in a group and waits until all pods reach the same stage of the scheduling/binding cycle before allowing any pods from the group to advance past that point.  If not all pods can reach that point before a timeout expires, then the scheduler stops trying to schedule that group, and all pods release all their resources.  This allows other workloads to try to allocate those resources.

Reviewer 2: @wojtek-t consider expanding the summary section to allow readers distinguishing the gang from workload concepts.

```yaml
apiVersion: v1
kind: Pod
spec:
  ...
  workload:
    name: wkld-1
  ...
```

The above pod might be one of several pods created by a `Job` like this.
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-1
spec:
  completions: 100
  parallelism: 100
  template:
    spec:
      restartPolicy: OnFailure
      containers:
      - name: ml-worker
        image: awesome-training-program:v1 
        command: ["python", "train.py"]
        resources:
          limits:
            nvidia.com/gpu: 1
        env:
        - name: JOB_COMPLETION_INDEX
          valueFrom:
            fieldRef:
              fieldPath:
               "metadata.annotations['batch.kubernetes.io/job-completion-index']"
```

The `Workload` core resource will be introduced. A `Workload` does not create any pods. It just describes what pods the scheduler should expect to see, and how to treat them.   

 It does not affect pod creation by Job or any other controller.  It only  A sample resource looks like this:
```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:
  name: myjob
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "pg1"
      gangMode: Single
      minCount: 100

>       schedulingTimeoutSeconds: 60

Reviewer 4: Yes. Thank you. There's also the current example that still have both fields on the `podGroup` level:
```yaml
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "pg1"
      gangMode: Single
      minCount: 100
      schedulingTimeoutSeconds: 60
```
->
```yaml
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "pg1"
      gangMode: Single
      gangSchedulingPolicy:
         minCount: 100
         schedulingTimeoutSeconds: 60
```

In these cases, there needs to be additional information to indicate which group a pod belongs to.

<<[UNRESOLVED @erictune @wojtek-t]>> 
Two options are being considered:
  * 

 <<[/UNRESOLVED]>>
The example below uses  `PodGroupSelector` on each `PodGroup` to identify pods.  For the Job example above, this looks like:

```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:
  name: myjobset
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "job1"
      gangMode: Single
      minCount: 100
      schedulingTimeoutSeconds: 60
      podGroupSelector:
        jobset.sigs.k8s.io/replicatedjob-name: job1
     - name: "job2"
      gangMode: Single
      minCount: 100
      schedulingTimeoutSeconds: 60
      podGroupSelector:
        jobset.sigs.k8s.io/replicatedjob-name: job2

```
This works well with existing controllers without changes.

Another option being considered is to specify the name of the `PodGroup` in the Pod itself, like this:

```yaml
apiVersion: v1
kind: Pod
name:
  jobset-job1-abc123
spec:
  ...
  workload:
    name: wkld-1
    podGroup: job1
  ...

> ```

Author: We can always encode int into string, so switching to string makes sense to me.
Switched.

When user downgrades the cluster to the version that no longer supports these two features:
- the `Workload` objects can no longer be created (the existing ones are not removed though)
- the `spec.workload` field can no longer be set on the Pods (the already set fields continue
  to be set though)
- scheduler reverts to the original behavior of scheduling one pod at a time ignoring
  existence of `Workload` objects and pods being linked to them

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> ## Production Readiness Review Questionnaire

Author: Still need to fill in some parts, but will not get to it today.

  <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.

### Associating Pod into PodGroups

When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
In these cases, there needs to be additional information to indicate which group a pod belongs to.

We proposed to extend the newly introduced `pod.spec.workload` field with additional information
to include that information. More specifically, the `pod.spec.workload` field is of type `PodWorkload`
and is defined as following:

```go
type WorkloadReference struct {
    // Workload defines the name of the Workload object this pod belongs to.
    Workload string

    // PodGroup defines the name of the PodGroup within a Workload this pod belongs to.
    PodGroup string

>     // PodGroupReplica defines the replica number within a PodGroup for PodGroups

Author: Yes - it's "which replica I belong to".

I'm definitely open to any other naming - I just think this will effectively be figured out during API review.

    // in ReplicatedGangMode that this pod belongs to.
    PodGroupReplica string
}
```

The example below shows how this could look like for with the following `Workload` object:

```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:
  name: jobset
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "job-1"

>       gangMode: Replicated

Reviewer 6: Does the name "gang" convey the meaning? I'd like to hear other people's opinions. I think it would be fine to unify it as workload.

## Proposal

The `spec.workload` field will be added to the Pod resource.  A sample pod with this new field looks like this:
```yaml
apiVersion: v1
kind: Pod
spec:
  ...
  workload:
    name: job-1
  ...
```

The above pod might be one of several pods created by a `Job` like this.
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-1
spec:
  completions: 100
  parallelism: 100
  template:
    spec:
      restartPolicy: OnFailure
      containers:
      - name: ml-worker
        image: awesome-training-program:v1 
        command: ["python", "train.py"]
        resources:
          limits:
            nvidia.com/gpu: 1
        env:
        - name: JOB_COMPLETION_INDEX
          valueFrom:
            fieldRef:
              fieldPath:
               "metadata.annotations['batch.kubernetes.io/job-completion-index']"
```

The `Workload` core resource will be introduced. A `Workload` does not create any pods. It just describes what pods the scheduler should expect to see, and how to treat them.   

 It does not affect pod creation by Job or any other controller.  A sample resource looks like this:
```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:

>   name: job-1

Author: This is great point.
It definitely is namespaced resource - the primary question is do we allow cross-namespace linking (i.e. workload being in a different namespace than that pod).

My assumption was "not" - and then we don't need namespace.
The question is do we agree on that...

<<[UNRESOLVED @erictune @wojtek-t]>> 
Two options are being considered:
  * 

 <<[/UNRESOLVED]>>
The example below uses  `PodGroupSelector` on each `PodGroup` to identify pods.  For the Job example above, this looks like:

```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:
  name: jobset
spec:
  podGroups:   # or gangGroups -- TBD
    - name: "job-1"
      gangMode: Single
      gangSchedulingPolicy:
        minCount: 100
        schedulingTimeoutSeconds: 60
      podGroupSelector:
        jobset.sigs.k8s.io/replicatedjob-name: job-1
    - name: "job-2"
      gangMode: Single
      gangSchedulingPolicy:
        minCount: 100
        schedulingTimeoutSeconds: 60
      podGroupSelector:
        jobset.sigs.k8s.io/replicatedjob-name: job-2

```
This works well for the simplest usecases, but becomes more complex with replicated gangs.
As an example, for `LeaderWorkerSet`, while pods have `leaderworkerset.sigs.k8s.io/group-index`
label, the actual podGroupSelector would become:
```
      podGroupSelector:
        leaderworkerset.sigs.k8s.io/group-index: <replica_number>
```

> Another option being considered is to specify the exact `PodGroup` and its replica number in the Pod

Reviewer: I prefer this because the first one would require some additional complexity like how we should tell to users if pod-a is specifying workload-a, but workload-a doesn't have a group that has matching selector.

Author: It is aligned with others preferences (including mine) - so let me already write is as a leading option and just mention the other as an alternative.

user intention was to have either both of them or none of them running, they should form a single group and
not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
may consist of a number of such replicas.

### Scheduler Changes

The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
to and from their `Workload` objects.

In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
the intention from the desired state.
Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`
object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

The north star vision for gang scheduling implementation should satisfy the following requirements:

> 1. Ensure that pods being part of a gang are not scheduled if all pods belonging to it can't be scheduled.

Author: done

2. Provide the "optimal enough" placement by considering all pods from a gang together.
3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.
4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different
   schedulers.
5. Avoid preemptions of already running pods if these will not allow to schedule higher priority gang
   anyway.

> 6. Support gang-level (or workload-level in general) level preemption.

Reviewer: Is each group (in workload.PodGroups[*]) supposed to have pods with the same template? Or can a single group contain various pods. I guess the latter is true, but in that case, the gang-aware preemption might be tricker because each pod in the same pod group could have a different priority.

Author: Yes - we definitely want to allow for the later (pods with different templates within a single PodGroup).

That being said, I would say that it doesn't mean that these can really be arbitrary pods with different priorities. We should discuss how to achieve that in the best way (whether you need to define priority explicitly at PodGroup level, whether we infer it somehow from the pods (min? max?) or something else, but having a gang of pods that would have different priorities shouldn't be supported usecase.

Author: My personal opinion is that we need a new "preemptionPolicy" entity that will effectively mean that certain groups of pods should be treated as a whole for preemption purposes. That will also contain the way how priority of it should be computed. This makes is strictly additive feature and should address the concerns.

That being said I added a paragraph to describe that and added a beta graduation criteria around that.

> 6. Support gang-level (or workload-level in general) level preemption.

Reviewer: Can we clarify a bit more? When I say "gang-aware preemption", I mean all pods from the group should be schedulable after the preemption. i.e., if pod#1-pod#15 can be schedulable after the preemption, but if the preemption cannot help making pod#16 schedulable, then all preemption shouldn't be performed.

I'd like to make sure we're on the same page, or there could be more basic requirements.

Author: > When I say "gang-aware preemption", I mean all pods from the group should be schedulable after the preemption. i.e., if pod#1-pod#15 can be schedulable after the preemption, but if the preemption cannot help making pod#16 schedulable, then all preemption shouldn't be performed.

Right - that's exactly what (5) was supposed to say. Hopefully the clarification suggested by Dominik helps here.

Regarding (6) - this is effectively a supplement of (5):
- in (5) we're saying that we won't preempt something, if we won't be schedule a gang anyway
- in (6) we're saying that if pods form a gang also from a runtime perspective, then we won't be preempting individual pods, but rather will be thinking about these as units

Reviewer: > in (6) we're saying that if pods form a gang also from a runtime perspective, then we won't be preempting individual pods, but rather will be thinking about these as units

Ok - please elaborate the explanation then. It's not very clear now

Author: Expanded the point (6) - PTAL

> 6. Support gang-level (or workload-level in general) level preemption.

Reviewer: So, (6) won't be supported even by the beta release.
What would be potential concerns not supporting it but shipping this gang feature to public? 

Let's say pod#16 cannot be helped by the preemption, while pod#1-pod#15 can. One potential bad scenario is that those pod#1-pod#15 keep preempting other pods while waiting for pod#16. That also would complicate the deadlock problem.  ...and the deadlock problem has to be solved by this beta release.

Author: See my comment above.

(6) is about different scenario - it's more about "preempting gangs as a whole" (i.e if a pod is part of a gang also from a runtime perspective (we will need additional "preemptionPolicy"), we can't preempt individual pods - we need to preempt gangs as a whole then.

Author: I updated the proposal as the way it was written was a bit misleading.
The requirement (5) to significant extent will generally be addressed in Beta release, because we won't start preemptions before having placement for all pods (the new Workload phase).
I added a paragraph explaining that.

Regarding gang-aware preemption (6) - I think it highly depends on what exactly we will do. Inferring gang priority from pods may indeed break users, but introducing new "preemptionPolicy" that will make the pods to be treated as gangs for preemption is an additive thing. This is also what I'm currently thinking.
So I'm fairly sure that we can actually solve it in backward-compatible way, but I high-level agree with your concern. For that part, I added the first Beta graduation criteria - having a high-level design for gang-preemption to ensure that we won't break compatibility with it.

PTAL

Reviewer 2: Do I understand correctly that the discussion is in fact when do we want to have the functionality become enabled by default (beta) vs staying disabled (in alpha)? Indeed having it enabled by default from the very beginning is tempting and we theoretically could try to make it happen with some restrictions like @sanposhiho said.

My thinking though was that we can start with alpha and target beta in the next release cycle.

- For me the problem of premature preemption (5) is a blocker for beta, because I can't imagine that gang physically preempts during its scheduling and before it is accepted.
- Preempting gangs as whole seems to me like a separate feature that can live in its own KEP and I wouldn't make it a blocker as we won't degrade the existing behavior.
- I also think we should address the problem of deadlocks before graduating to beta.

Author: +1 - this is aligned with my thinking

And I think this is also aligned with what is described in the KEP now - first and third bullets are things that we have a plan.
I agree the second bullet is a separate feature, and I don't want to block Beta graduation on having it implemented. But I think that the concern that @sanposhiho is a valid one and ensuring that we have a backward-compatible path forward is an important one. I believe that the high-level approach I added is this mitigation but ensuring that we all agree on that should be a beta blocker.
So it's not about having that feature working - it's about agreeing that we believe in feasibility of that approach - that is what I believe should be a beta blocker.

Reviewer: Now that (5) is added to a beta requirement, I have no concern at all here.

So, you're basically saying gang pods shouldn't go through the existing pod-level preemption in the scheduling cycle and instead we expect a new preemption phase to be newly implemented at the workload level, right?

Re: breaking change, great that we have it at the beta requirement so that we will surely revisit this topic at the next release cycle.

> 

Reviewer 2: 7. Updating workload status and triggering rescheduling when a gang failed binding in the all-or-nothing fashion.

Author: done

While we will not target addressing "optimal enough" part of requirement (2), we will prove that we
can process all pods together. The single scheduling cycle and blocking resources will address the
requirement (3).

More detail about scheduler changes is described in [this document](https://docs.google.com/document/d/1lMYkDuGqEoZWfE2b8vjQx0vHieOMyfmi6VHUef5-5is/edit?tab=t.0#heading=h.1p88ilpefnb).

We will continue with further improvements on top of it with follow-up KEPs. We are planning to
introduce the concept of `Reservation` that will allow to treat distributed subset of resources as
a single unit from scheduling perspective. With that, the proposed placement being a result of
the scheduling decision of the `Workload` phase will become a `Reservation`. This will become the
coordination point and a mechanism for multiple schedulers to share the underlying infrastructure
addressing the requirement (4). This will also be a critical building block for workload-level
preemption and addressing requirement (6). Further extensions to `Reservation` with different states
(e.g. not yet block resources) will help with addressing the unnessary preemptions issue -

> requirement (5).

Reviewer 2: ```suggestion
requirement (5). Finally making the binding process aware of gangs will allow to make sure the process is either successful or triggers workload rescheduling satisfying (7) requirement.
```

Author: done

have been executed regularly, and have been stable.
This can be done with:
- permalinks to the GitHub source code
- links to the periodic job (typically a job owned by the SIG responsible for the feature), filtered by the test name
- a search in the Kubernetes bug triage tool (https://storage.googleapis.com/k8s-triage/index.html)

We expect no non-infra related flakes in the last month as a GA graduation criteria.
If e2e tests are not necessary or useful, explain why.
-->

- [test name](https://github.com/kubernetes/kubernetes/blob/2334b8469e1983c525c0c6382125710093a25883/test/e2e/...): [SIG ...](https://testgrid.k8s.io/sig-...?include-filter-by-regex=MyCoolFeature), [triage search](https://storage.googleapis.com/k8s-triage/index.html?test=MyCoolFeature)

### Graduation Criteria

#### Alpha

- Workload API is introduced behind Workload feature flag
- API tests for Workload API (that will be promoted to conformance in GA release)
- kube-scheduler implements first version of gang-scheduling based on groups defined in the Workload object

#### Beta

- TBD in for Beta release

#### GA

- TBD in for Beta release

### Upgrade / Downgrade Strategy

> This KEP effectively boils down to two separate functionalities:

Author: done

- the Workload API and new field in Pod API that allows linking Pods to Workloads
- scheduler changes implementing the gang scheduling functionality

When user upgrades the cluster to the version that supports these two features:
- they can start using the new API by creating Workload objects and linking pods to it via
  explicitly specifying their new `spec.workload` field
- scheduler automatically uses the new extensions and tries to schedule all pods from a given
  gang in a scheduling group based on the defined `Workload` objects

When user downgrades the cluster to the version that no longer supports these two features:
- the `Workload` objects can no longer be created (the existing ones are not removed though)
- the `spec.workload` field can no longer be set on the Pods (the already set fields continue
  to be set though)
- scheduler reverts to the original behavior of scheduling one pod at a time ignoring
  existence of `Workload` objects and pods being linked to them

### Version Skew Strategy

> The feature is limited to control plane, so the version skew with nodes (kubelets) doesn't matter.

Author: done

Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

- [X] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: Workload

Author: > Why do we need two feature gates?

My rationale for that is that if we find an issue in the gang-scheduling feature, we shouldn't be forced to disable the API changes in kube-apiserver - just disabled the feature itself.

I'm definitely opened for other names - added suggestions.

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [ ] Metrics
  - Metric name:
  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

No dependendies other than the components where the feature is implemented
(kube-apiserver and kube-scheduler).

### Scalability

###### Will enabling / using this feature result in any new API calls?

Yes:

Watching for workloads:
  - API call type: LIST+WATCH Workloads
  - estimated throughput: < XX/s
  - originating component: kube-scheduler, kube-controller-manager (GC)

Status updates:

>   - API call type: PUT/PATCH Workloads

Author: Added.

The `Workload` object will allow kube-scheduler to be aware that pods are part of workloads with complex internal structure.  Those workloads include builtins like `Job` and `StatefulSet`, and custom workloads, like `JobSet`, `LeaderWorkerSet`, `MPIJob` and `TrainJob`. All of these workload types are used for AI training and inference use cases.

### Goals
- Introduce a concept of a `Workload` as a primary building block for workload-aware scheduling vision
- Implement the first version of `Workload` API necessary for defining a Gang
- Ensuring that we can extend `Workload` API in backward compatible way toward north-star API
- Ensuring that `Workload` API will be usable for both built-in and third-party workload controllers and APIs

> - Implement first version of gang-scheduling in kube-scheduler

Reviewer 7: Can we provide the name of main functionalinies (i.e. WaitOnPermit)  that we expect to be implemented in the first version?

- Provide full backward compatibility for all existing scheduling features

### Non-Goals

- Take away responsibility to create pods from controllers.
- Bring fairness or multiple workload queues in kube-scheduler. Kueue and Volcano.sh will continue to provide this.
- Map all the declarative state and behaviors into `Workload` object. It is focused only on scheduling-related parts.

The following are non-goals for this KEP but will probably soon appear to be goals for follow-up KEPs:

- Introduce a concept of `Reservation` that can be later consumed by pods.
- Workload-level preemption.
- Address resource contention between different schedulers (including possible deadlocks).
- Address the problem of premature preemptions in case the higher priority workloads does not

>   eventually schedule.

Reviewer 7: Why do we need to list the future goals? If we want to do so, should we move them from non-goals to goals section?

## Proposal

The `spec.workload` field will be added to the Pod resource.  A sample pod with this new field looks like this:
```yaml
apiVersion: v1
kind: Pod
spec:
  ...
  workload:
    name: job-1
  ...
```

The above pod might be one of several pods created by a `Job` like this.
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-1
spec:
  completions: 100

>   parallelism: 100

Reviewer 8: For sake of correctness, shouldn't this Job has its completionMode set ?

```suggestion
  parallelism: 100
  completionMode: Indexed
```

  template:

>     spec:

Reviewer 8: If my understanding is correct we don't be touching the Job controller at this stage so the user would need to also provide the Workload as follow, no ?

```suggestion
    spec:
      workload:
        name: job-1
```

* In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called
  <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.

### Associating Pod into PodGroups

When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
In these cases, there needs to be additional information to indicate which group a pod belongs to.

We proposed to extend the newly introduced `pod.spec.workload` field with additional information
to include that information. More specifically, the `pod.spec.workload` field is of type `PodWorkload`
and is defined as following:

```go

> type WorkloadReference struct {

Reviewer 7: ```suggestion
// WorkloadReference identifies the Workload object and gang membership
// (PodGroup and, optionally, its replica) that a Pod belongs to.
// The scheduler uses this information to enforce gang semantics, and
// controllers can use it to correlate Pod state with workload/gang status.
type WorkloadReference struct {
```

    // Workload defines the name of the Workload object this pod belongs to.

>     Workload string

Reviewer 7: I prefer to rename it to either `workloadName` or just `name`. Also, should we mention that it is namespaced "same namespace as the Pod"?

    // PodGroup defines the name of the PodGroup within a Workload this pod belongs to.

>     PodGroup string

Reviewer 7: What is the use/importnace of this field? Would the `PodGroupReplica` field is enough?

spec:
  ...
  workload:
    name: jobset
    podGroup: job-1
    podGroupReplica: 2
  ...

```

We decided for this option because it is more succint and makes the role of a pod clear just
from inspecting the pod (and simple/efficient to group).
We acknowledge the fact that this option may require additional minor changes in the controllers
to adopt this pattern (e.g. for LeaderWorkerSet we will need to populate the pod template
similarly that we currently populate the labels).

The primary alternative we consider was to introduce the the `PodGroupSelector` on each `PodGroup`
to identify pods belonging to it. However, with this pattern:
- there are additional corner cases (e.g. a pod links to a workload but none of its PodGroups matching
  that pod)
- for replicated gang, we can't use the full label selector, but rather support specifying only the
  label key, similar to `MatchLabelKeys` in pod affinity

### API

The `Workload` type will be defined with the following structure:
```go
type Workload struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	Spec WorkloadSpec
	Status WorkloadStatus
}

type ReplicaMode string

const (
    ReplicaModeUnreplicated ReplicaMode = "Unreplicated"

> )

Reviewer 7: What is the use of this type?

// WorkloadSpec describes a workload in a portable way that scheduler and related
// tools can understand.  
type WorkloadSpec struct {
    // ControllerRef points to the true workload, e.g. Deployment.
    // It is optional to set and is intended to make this mapping easier for
    // things like CLI tools.
    ControllerRef *v1.ObjectReference
    // PodGroups is a list of groups of pods.
    // Each group may request gang scheduling.
    PodGroups []PodGroup 
}

type GangMode string
const (

> 	// GangModeGang means that all pods in this PodGroup need to be scheduled as one gang.

Reviewer 2: ```suggestion
	// GangModeSingle means that all pods in this PodGroup need to be scheduled as one gang.
```

> 	GangModeSingle GangMode = "Single"

Reviewer 7: Can't we use the replica field to detirmine if it is single or not? So:

- **GangMode**: `off` or` on` (default). No “ReplicatedGangMode” as a separate value.
- **Replica**: integer, defaults to 1.
	- If Replica == 1 → the PodGroup is a single gang.
	- If Replica > 1 → the PodGroup is replicated, each replica gets its own index.
	
	wdyt?

Reviewer 2: +1.
We can have replicated gangs with the GangMode `off`?

// multiple dense indexes (RankedGroups) and which can optionally be replicated in a variable
// number of identical copies.
//
// TODO: Decide on the naming: PodGroup vs GangGroup.
type PodGroup struct {
    Name *string
    GangMode *GangMode // default is "Single"

    // Optional when GangMode = "ReplicatedGang".
    // Forbidden otherwise.
    Replicas int

    // GangSchedulingPolicy defines the options applying to all pods in this gang.
    // Forbidden if GangMode is set to "Off".
    GangSchedulingPolicy GangSchedulingPolicy
}

type WorkloadStatus struct {
  // Necessary status fields TBD.
}
```

The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
user intention was to have either both of them or none of them running, they should form a single group and
not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
may consist of a number of such replicas.

> ### Scheduler Changes

Reviewer 7: I'd add a dedicated section for Admission Semantics that elaborate on how the gang scheduling will change the addmision flow. (reservasion, rollback, timeout, retry).

-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No.

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### How can an operator determine if the feature is in use by workloads?

<!--
Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

###### How can someone using this feature know that it is working for their instance?

<!--
For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
for each individual pod.
Pick one more of these and delete the rest.
Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

> - [ ] Events

Reviewer 7: Can you list the events and status?

### Associating Pod into PodGroups

When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
In these cases, there needs to be additional information to indicate which group a pod belongs to.

We proposed to extend the newly introduced `pod.spec.workload` field with additional information
to include that information. More specifically, the `pod.spec.workload` field is of type `PodWorkload`
and is defined as following:

```go
type WorkloadReference struct {
    // Workload defines the name of the Workload object this pod belongs to.
    Workload string

    // PodGroup defines the name of the PodGroup within a Workload this pod belongs to.
    PodGroup string
    // PodGroupReplica defines the replica number within a PodGroup for PodGroups
    // in ReplicatedGangMode that this pod belongs to.

>     PodGroupReplica string

Author: Yeah - I switched to string because there are controllers using non-number for that and I think we want to support it (and we can encode int as string too).

	GangModeReplicated GangMode = "Replicated"
)

// GangSchedulingPolicy holds options that affect how gang scheduling of one PodGroup is handled by the scheduler.
type GangSchedulingPolicy struct {
    // TODO: Decide between SchedulingTimeoutSeconds and making this field of type time.Duration.
	SchedulingTimeoutSeconds *int
	MinCount *int
}

// PodGroup is a group of pods that may contain multiple shapes (EqGroups) and may contain
// multiple dense indexes (RankedGroups) and which can optionally be replicated in a variable
// number of identical copies.
//
// TODO: Decide on the naming: PodGroup vs GangGroup.
type PodGroup struct {
    Name *string
    GangMode *GangMode // default is "Single"

	// ReplicatedGangKey must be set when GangMode = "Replicated"
    // It must be one of:
    //   - `metadata.labels['<KEY>']`
    //   Note: Annotations are not supported because they are not indexed easily, and may not
    //         have a form that can be printed without escaping.
    // Each value of that key maps to a different identical gang, which will be anonymous in status,
    // and referred to by the key's value in events and errors.

>     ReplicatedGangKey *ObjectFieldSelector

Author: Right - removed.

The following are non-goals for this KEP but will probably soon appear to be goals for follow-up KEPs:

- Introduce a concept of `Reservation` that can be later consumed by pods.
- Workload-level preemption.
- Address resource contention between different schedulers (including possible deadlocks).
- Address the problem of premature preemptions in case the higher priority workloads does not
  eventually schedule.

## Proposal

The `spec.workload` field will be added to the Pod resource.  A sample pod with this new field looks like this:
```yaml
apiVersion: v1
kind: Pod
spec:
  ...
  workload:
    name: job-1
  ...
```

The above pod might be one of several pods created by a `Job` like this.
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-1
spec:
  completions: 100
  parallelism: 100
  template:
    spec:
      restartPolicy: OnFailure
      containers:
      - name: ml-worker
        image: awesome-training-program:v1 
        command: ["python", "train.py"]
        resources:
          limits:
            nvidia.com/gpu: 1
        env:
        - name: JOB_COMPLETION_INDEX
          valueFrom:
            fieldRef:
              fieldPath:

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Author: That's actually great point - I don't see any good reason of not doing that.

@erictune @dom4ha - WDYT?

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Reviewer 7: I agree with @soltysh , if we are going to add workload ref in the pod spec, there is no need to add it in the Job spec. 

What we have discussed in the design proposal doc was to add ref to the object in the workload spec and ref for the workload on each object.

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Reviewer 7: @soltysh @wojtek-t , After second thought and discussing this with @erictune, I had different thoughts.

Having a workload field directly in `pod.spec` would indeed let Job-stamped Pods carry the workload identity automatically. But we still need `workloadRef` on the object resource (in my KEP case, Job) itself for a couple of reasons:

-  If controllers try to “walk Pods” to infer the workload, they don’t know when to stop. That becomes ambiguous for retries and hierarchical controllers (i.e. JobSet → Job → Pods).
- The Job controller owns retries, completions, and status. If only Pods declare the workload, the Job would have to infer gang outcomes by inspecting children. With an explicit `job.spec.workloadRef`, the controller can surface clear conditions without guesswork.
- Other controllers (JobSet, LWS, RayJob, etc.) also need this. Having a top-level reference avoids each one re-implementing inference logic, which gives us cross-controller consistency.
- A Job level field can be validated up front by API server, whereas a Pod-only field appears only after creation.

So in practice we need both: Pods carry the reference for the scheduler, and the Job declares it as the authoritative source of truth. That’s the gap KEP #5548 is addressing.

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Reviewer 3: @wojtek-t  This is  not strictly required for alpha, but I think it will improve UX.

Imagine a user who is debugging scheduler behavior. They look at scheduler logs or events.  They see scheduler events about Workload X preempting Workload Y.  Then they want to drill into the specifics of Workload X and Workload Y (e.g. that X is a Job and what its name is, and that Y is a MPIJob and what its name is, so they can look these up.)  This could be after pods have terminated and been GC'ed.  The ownerReference information is missing.  And if it is present, it is tedious to traverse.

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Author: @erictune - I think you're referring to a different thing. There are two things:

1) Having `ControllerRef` inside Workload object (part of this KEP)
2) Adding `workloadRef` in Job API (part of https://github.com/kubernetes/enhancements/pull/5548 - @helayoty KEP)

Your comment above shows why (1) makes sense. I think it's not critical, but I also see a value in it. But that's not the thing that is being questioned in this comment thread.

What is being questioned here is (2). Because as pointed by @soltysh (which I agreed with), introducing `workloadRef` in Job API is a duplication of information. Because the exact same thing can already be expressed through the `WorkloadReference` in the `PodTemplateSpec`.

So we will already have a way to get the information without the #5548:
(a) From Workload to Job - using the `ControllerRef` field being part of WorkloadSpec
(b) From Job to Workload - using the `job.spec.podTemplate.WorkloadReference` field

>                "metadata.annotations['batch.kubernetes.io/job-completion-index']"

Author: The discussion diverged into two places, so cross-referencing the summary for  (2) here:
https://github.com/kubernetes/enhancements/pull/5558/files#r2410167467

```

The `Workload` core resource will be introduced. A `Workload` does not create any pods. It just describes what pods the scheduler should expect to see, and how to treat them.   

 It does not affect pod creation by Job or any other controller.  A sample resource looks like this:
```yaml
apiVersion: scheduling/v1alpha1   
kind: Workload
metadata:
  namespaces: ns-1
  name: job-1
spec:

>   podGroups:   # or gangGroups -- TBD

Author: PodSet (as described in the naming) will be a lower layer within a PodGroup.

GangGroup no longer makes sense if we explicitly support non-gang. So for now we're sticking with PodGroup.

    - name: "pg1"
      gangMode: Single

>       gangSchedulingPolicy:

Reviewer 9: Yeah, I figure you will. I'll put my api-reviewer shadow hat on and will propose it again then :wink:

>       gangSchedulingPolicy:

Author: I went ahead and updated it further given we updated the proposal here.

* In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called
  <<[UNRESOLVED depending on a previous unresolved item]>> `PodSet` if `PodGroup` is chosen, or else `EqGroup` if `GangGroup` is chosen<<[/UNRESOLVED]>>.

### Associating Pod into PodGroups

When a `Workload` consists of a single group of pods needing Gang Scheduling, it is clear which pods belong to the group from the `spec.workload.name` field of the pod.  However `Workload` supports listing multiple list items, and a list item can represent a single group, or a set of identical replica groups.
In these cases, there needs to be additional information to indicate which group a pod belongs to.

We proposed to extend the newly introduced `pod.spec.workload` field with additional information
to include that information. More specifically, the `pod.spec.workload` field is of type `PodWorkload`
and is defined as following:

```go

> type WorkloadReference struct {

Reviewer: Re: 2 - So, you mean... it should be rejected, right? like PreEnqueue stage will block those pods until the workload is created. And I meant, then, a error report will depend on https://github.com/kubernetes/enhancements/pull/5510.

> type WorkloadReference struct {

Reviewer 2: rejected = blocked until workload object appears + error reporting

Typically workload object should be created before pods, but we don't have to rely on it.

}

type WorkloadStatus struct {
  // Necessary status fields TBD.
}
```

The individual `PodGroups` and `PodGroup` replicas are treated as independent gangs. As an example, if one of
the groups can be scheduled and the other can't be - this is exactly what will happen. If the underlying
user intention was to have either both of them or none of them running, they should form a single group and
not be split into two. A `LeaderWorkerSet` is a good example of it, where a single `PodGroup` replica consists
of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
may consist of a number of such replicas.

### Scheduler Changes

The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
to and from their `Workload` objects.

In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
the intention from the desired state.
Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 7: >Add these fields to Job to allow overriding defaults:
>
>spec.gangMode defaults to on for indexedJob and off for others, but user can override by setting.
>spec.gangSchedulingTimeoutSeconds defaults to controller-wide default, but user can override by setting.

Does that mean that you want to add the following fields to the Job API?
- `spec.workloadRef`
- `spec.gangMode`
- `spec.gangSchedulingTimeoutSeconds`

If that is the case, do you think we are replicating many info in W into Job API? What is the case when W is getting extended with more setting? 

As an alternative, how about creating/linking W if the `spec.workloadRef.Name` is not empty?

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 9: I'm still having a hard time with these ideas. If you control the whole Job object, why would you want to differentiate between those two fields (one at job spec level, and the other at pod template level) which as an author of a job have a full control of? What's the value of duplicating this information?

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 9: I'm inclined to push this such that for alpha, you only leave the initial part talking about user-managed Workload resource.

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 9: Also, in the long run, does that mean we'll be extending all of built-in controllers with similar changes? I'm having a hard time supporting this kind of additions across the board.

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 3: @helayoty -- I like the suggestion of putting `spec.gangMode` and `spec.gangSchedulingTimeoutSeconds` somewhere in Job Spec.  I don't consider it a problem that some information is replicated. The user only fills in Job, so from their standpoint, it is not replicated.

I don't think adding `spec.workloadRef` to Job is that useful.  The same information can be put into the Pod template for Job.  

For some other types of workload, which do not expose the pod template to the user, such as  `TrainJob`, then it may be necessary to expose `spec.workloadRef` in those true controllers.  But not for Job.

@soltysh, yes, all built-in controllers and many custom controllers will get some fields added.   However, the exact set of fields added will be different per controller.   Different controllers have purposes and design philosophies, so one size does not fit all.

For example Deployments do not typically need gang scheduling, so I expect that we don't ever add `spec.gangMode` to Deployment.  But, we may later develop a new version of topology spreading and define that inside Workload.  In that case, we would add a `spec.spreading` stanza to Deployment, and the Deployment controller would copy that to`Workload.spec.podgroup[].topologySpreading`.

> object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

Reviewer 7: I'm ok with (c).

will result in pods no longer being treated individually for preemption purposes) which makes it
an additive feature. However, having a next level of details is required to ensure that we really
have a feasible backward-compatible plan before promoting this feature to Beta.

However, approval for this KEP is NOT an approval for this vision. We only sketch it to show that
we see a viable path forward from the proposed design that will not require significant rework.

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

N/A

> ##### Unit tests

Reviewer 9: Can you update this section with the packages you're planning to touch?

    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--

> Why should this KEP _not_ be implemented?

Author: Added

The `Workload` object will allow kube-scheduler to be aware that pods are part of workloads with complex internal structure.  Those workloads include builtins like `Job` and `StatefulSet`, and custom workloads, like `JobSet`, `LeaderWorkerSet`, `MPIJob` and `TrainJob`. All of these workload types are used for AI training and inference use cases.

### Goals
- Introduce a concept of a `Workload` as a primary building block for workload-aware scheduling vision
- Implement the first version of `Workload` API necessary for defining a Gang
- Ensuring that we can extend `Workload` API in backward compatible way toward north-star API
- Ensuring that `Workload` API will be usable for both built-in and third-party workload controllers and APIs

> - Implement first version of gang-scheduling in kube-scheduler

Reviewer 10: What scheduling constraints will or will not be supported? pod affinity/anti-affinity, node affinity/selector, pod topology spread? if all of them, then I would mention that explicitly. 

@erictune

- Provide full backward compatibility for all existing scheduling features

> ### Non-Goals

Reviewer 10: Cluster autoscaling support or lack of is completely absent in this KEP, I believe it should be explicitly called out.

I have my own workload definition (CRD) and controller managing its lifecycle. I would
like to be able to easily benefit of gang-scheduling feature supported by the core
Kubernetes without extensive changes to my custom controller.

### Risks and Mitigations

#### The API needs to be extended in an unpredictable way

We try to mitigate it by an extensive analysis of usecases and already sketching
how we envision the direction in which the API will need to evolve to support further
usecases. You can read more about it in the [extended proposal] document.

## Design Details

### Naming

* `Workload` is the resource Kind.
* `scheduling` is the ApiGroup.
* `spec.workload` is the name of the new field in pod.
* Within a Workload there is a list of groups of pods. Each group represents a top-level division of pods within a Workload.  Each group can be independently gang scheduled (or not use gang scheduling). This group is named
* In a future , we expect that this group can optionally specify further subdivision into sub groups.  Each sub-group can have an index.  The indexes go from 0 to N, without repeats or gaps. These subgroups are called

> * In subsequent KEPs, we expect that a sub-group can optionally specify further subdivision into pod equivalence classes.  All pods in a pod equivalence class have the same values for all fields that affect scheduling feasibility.  These pod equivalence classes are called

Reviewer 11: You accidentally removed the names for these three

    PodGroups []PodGroup 
}

// PodGroup is a group of pods that may contain multiple shapes (PodSets) and may contain
// multiple dense indexes (PodSubGroups) and which can optionally be replicated in a variable
// number of identical copies.
type PodGroup struct {
    Name *string

    // Number of identical instances of PodGroup that are part of the Workload.
    // Defaults to 1.
    Replicas int

    // Policy defines the configuration of the PodGroup to enable different
    // scheduling policies.
    Policy PodGroupPolicy
}

// PodGroupPolicy defines scheduling configuration of a PodGroup.
type PodGroupPolicy struct {
    // Exactly one of the policies should be set.

    // Default scheduling policy (default Kubernetes behavior).
    Default *DefaultSchedulingPolicy

    // Gang scheduling policy (all-or-nothing scheduling semantics)
    Gang *GangSchedulingPolicy
}

// DefaultSchedulingPolicy represents default scheduling behavior.
type DefaultSchedulingPolicy struct {
    // For now this is effectively just a marker type.
}

// GangSchedulingPolicy represents options for how gang scheduling of one
// PodGroup should be handled.
type GangSchedulingPolicy struct {

>     MinCount *int

Reviewer 7: Why has `SchedulingTimeoutSeconds` field been removed?

Reviewer 3: I agree with the plan to move to scheduler config.

	metav1.TypeMeta
	metav1.ObjectMeta
	Spec WorkloadSpec
	Status WorkloadStatus
}

// WorkloadSpec describes a workload in a portable way that scheduler and related
// tools can understand.  
type WorkloadSpec struct {
    // ControllerRef points to the true workload, e.g. Deployment.
    // It is optional to set and is intended to make this mapping easier for
    // things like CLI tools.
    // This field is immutable.
    ControllerRef *v1.ObjectReference

    // PodGroups is a list of groups of pods.
    // Each group may request gang scheduling.
    PodGroups []PodGroup 
}

// PodGroup is a group of pods that may contain multiple shapes (PodSets) and may contain
// multiple dense indexes (PodSubGroups) and which can optionally be replicated in a variable
// number of identical copies.
type PodGroup struct {
    Name *string

    // Number of identical instances of PodGroup that are part of the Workload.
    // Defaults to 1.
    Replicas int

    // Policy defines the configuration of the PodGroup to enable different
    // scheduling policies.
    Policy PodGroupPolicy
}

// PodGroupPolicy defines scheduling configuration of a PodGroup.
type PodGroupPolicy struct {

>     // Exactly one of the policies should be set.

Reviewer 9: Nit: from api perspective you'll want to add a discriminator field, expressing which policy is at play. That's what we've been doing with all the union-like types.

of a single leader and `N` workers and that forms a scheduling (and runtime unit), but workload as a whole
may consist of a number of such replicas.

### Scheduler Changes

The kube-scheduler will be watching for `Workload` objects (using informers) and will use them to map pods
to and from their `Workload` objects.

In the initial implementation, we expect users to create the `Workload` objects. In the next steps controllers
will be updated to create an appropriate `Workload` objects themselves whenever they can appropriately infer
the intention from the desired state.
Note that given scheduling options are stored in the `Workload` object, pods linked to the `Workload`
object will not be scheduled until this `Workload` object is created and observed by the kube-scheduler.

The north star vision for gang scheduling implementation should satisfy the following requirements:

1. Ensure that pods being part of a gang are not bound if all pods belonging to it can't be scheduled.
2. Provide the "optimal enough" placement by considering all pods from a gang together.
3. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by kube-scheduler.

> 4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by different

Reviewer 9: Any particular reason you have this element mentioned twice? Why can't you just write this as:
```suggestion
4. Avoid deadlock scenario when multiple workloads are being scheduled at the same time by any scheduler (kube-scheduler, or third-party provided scheduler). 
```

However, approval for this KEP is NOT an approval for this vision. We only sketch it to show that
we see a viable path forward from the proposed design that will not require significant rework.

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

N/A

##### Unit tests

- `k8s.io/kubernetes/pkg/scheduler`: `2025-10-02` - 81.7%
- `k8s.io/kubernetes/pkg/scheduler/backend/queue`: `2025-10-02` - 91.4%
- `k8s.io/kubernetes/pkg/scheduler/framework`: `2025-10-02` - 81.7%
- `k8s.io/kubernetes/pkg/scheduler/framework/preemption`: `2025-10-02` - 64.2%

> - `k8s.io/kubernetes/pkg/scheduler/framework/util/assumecache`: `2025-10-02` - 86.2%

Reviewer 9: You'll probably want to include also:
```
- k8s.io/kubernetes/pkg/apis/scheduling/v1alpha1
- k8s.io/kubernetes/pkg/apis/scheduling/validation
```
since that's where `Workload` resource will live.

Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

- [X] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: Workload/GenericWorkload/NativeWorkload

Reviewer 9: `kep.yaml` talks about `Workload`, would be nice to pick one and use it consistently across all places :wink:

The feature should start working again.
However, the user need to remember that some Workload objects could already be stored
in etcd and may affect the behavior of some of the existing workloads.

###### Are there any tests for feature enablement/disablement?

No.
The enablement/disablement for the new field in Pod API will be added similarly to this PR:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282

Note that gang-scheduling itself is purely in-memory feature, so feature themselves are enough.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

> ###### What specific metrics should inform a rollback?

Reviewer 9: Although not required, have you considered what kind of metrics you could expose for this functionality and where? I'm thinking about exposing gang-related metrics in the scheduler, as the most appropriate place to track this.

Reviewer 9: I was thinking primarily about the gangscheduling feature, since like most of our scheduling code it's hard to follow from outside. I like both of the proposed metrics. Now I need to only remember that you wrote them here for beta :sweat_smile:

ignoring the rest of the requirements for `Alpha` phase.

For `Beta`, we want to also touch requirements (2) and (3) by extending the scheduling framework with
a new dedicated phase (tentatively called Workload). In that phase,
kube-scheduler will be looking at all pods from a gang (part of `Workload`) and compute the placement
for all of these pods in a single scheduling cycle. Those placements will be stored only in-memory and
block the required resources from scheduling. Tentively we plan to use `NominatedNodeName` field for it.
After that, pods will go through regular pod-by-pod scheduling phases (including Filter and Score)
with a nomination as a form of validation the proposed placement and execution of this placement decision.
Therefore we expect the order of processing pods won't ever be important, but all-or-nothing nature of
gangs will be preserved while advancing through the further steps of the binding process.

While we will not target addressing "optimal enough" part of requirement (2), we will assure that we
can process all gang pods together. The single scheduling cycle and blocking resources in beta
will address the requirement (3).

> We will also introduce delayed preemption by moving it after `WaitOnPermit` phase. Together with

Reviewer: What if a pod deletion takes long? The current scheduler mitigates it by just letting a preemptor pod go anywhere else (if there's an empty space, probably made by some other pod termination), while a deletion isn't completed.
